File: libraries/C/wasmbot_client.c
#define WASMBOTSMESSAGE_IMPLEMENTATION
#include "./wasmbot_client.h"

#include <stdarg.h>
#include <string.h>
#include <stdbool.h>
#include <stdio.h>


//// LOGGING
extern void logFunction(int logLevel, unsigned int msgPtr, unsigned int msgLen);

void wsmbt_log(const char* msg) {
    logFunction(2, (size_t)msg, strlen(msg));
}

void wsmbt_logErr(const char* msg) {
    logFunction(0, (size_t)msg, strlen(msg));
}

void wsmbt_logf(const char* fmt, ...) {
    va_list argptr;
    va_start(argptr, fmt);
    size_t buffSizeNeeded = vsnprintf(NULL, 0, fmt, argptr);
    va_end(argptr);
    char* msgBuffer = malloc(buffSizeNeeded+1);
    va_start(argptr, fmt);
    vsprintf(msgBuffer, fmt, argptr);
    va_end(argptr);

    wsmbt_log(msgBuffer);
    free(msgBuffer);
}

void wsmbt_logfErr(const char* fmt, ...) {
    va_list argptr;
    va_start(argptr, fmt);
    size_t buffSizeNeeded = vsnprintf(NULL, 0, fmt, argptr);
    va_end(argptr);
    char* msgBuffer = malloc(buffSizeNeeded+1);
    va_start(argptr, fmt);
    vsprintf(msgBuffer, fmt, argptr);
    va_end(argptr);

    wsmbt_logErr(msgBuffer);
    free(msgBuffer);
}
//// \LOGGING


//// RANDOM NUMBERS
extern int32_t getRandomInt(int32_t min, int32_t max);
int32_t wsmbt_getRandomInt(int32_t min, int32_t max) {
    return getRandomInt(min, max);
}
//// \RANDOM NUMBERS


//// MEMORY CREATION + READING AND WRITING
uint8_t* WSMBT_HOST_RESERVE = NULL;
size_t WSMBT_HOST_RESERVE_SIZE = 0;

bool _reserveMemory(size_t request) {
    if (WSMBT_HOST_RESERVE != NULL) {
        wsmbt_logErr("CLIENT ERROR: Attempting to reserve memory twice");
        return false;
    }
    WSMBT_HOST_RESERVE = malloc(request);
    if (WSMBT_HOST_RESERVE == NULL) {
        wsmbt_logErr("CLIENT ERROR: Could not allocate reserve memory");
        return false;
    }
    WSMBT_HOST_RESERVE_SIZE = request;
    return true;
}
//// \MEMORY CREATION + READING AND WRITING


//// GAME PARAMETERS
const uint16_t GP_VERSION = 7;

extern wsmbt_BotMetadata clientSetup(void);
extern bool clientReceiveGameParams(WasmBotsMessage_InitialParameters*);

bool receiveGameParams(size_t offset) {
    wsmbt_GameParameters params;

    WasmBotsMessage_DataAccess da = {
        .buffer = WSMBT_HOST_RESERVE,
        .bufferSize = WSMBT_HOST_RESERVE_SIZE,
        .position = offset,
    };

    WasmBotsMessage_InitialParameters initParams;
    WasmBotsMessage_InitialParameters_FromBytes(&da, &initParams);

    if (initParams.paramsVersion != GP_VERSION) {
        wsmbt_logfErr("CLIENT ERROR: Can't parse GameParams v%d; only prepared for v%d", params.paramsVersion, GP_VERSION);
        return false;
    }

    return clientReceiveGameParams(&initParams);
}
//// \GAME PARAMETERS


//// SETUP AND LOOP
WasmBotsMessage_DataAccess _dataAccess;

size_t setup(size_t requestReserve) {
    if (!_reserveMemory(requestReserve)) {
        return 0;
    }

    _dataAccess.buffer = WSMBT_HOST_RESERVE;
    _dataAccess.bufferSize = requestReserve;
    _dataAccess.position = 0;

    const wsmbt_BotMetadata botData = clientSetup();

    memcpy((char*)(WSMBT_HOST_RESERVE), botData.name, WSMBT_BOT_MAX_NAME_LEN);

    _dataAccess.position = WSMBT_BOT_MAX_NAME_LEN;
    WasmBotsMessage__WriteUInt16(&_dataAccess, botData.version[0]);
    WasmBotsMessage__WriteUInt16(&_dataAccess, botData.version[1]);
    WasmBotsMessage__WriteUInt16(&_dataAccess, botData.version[2]);

    return (size_t)WSMBT_HOST_RESERVE;
}

void* _noop(WasmBotsMessage_PresentCircumstances* circumstances) { return NULL; }
wsmbt_TickFunction _clientTick = &_noop;

void wsmbt_registerTickCallback(wsmbt_TickFunction tickFunc) {
    _clientTick = tickFunc;
}

void tick(size_t offset) {
    WasmBotsMessage_err_t err = WASMBOTSMESSAGE_ERR_OK;
    void* playerMove = NULL;

    _dataAccess.position = offset;
    WasmBotsMessage_PresentCircumstances* circumstances = malloc(sizeof(WasmBotsMessage_PresentCircumstances));
    err = WasmBotsMessage_PresentCircumstances_FromBytes(&_dataAccess, circumstances);
    if (err != WASMBOTSMESSAGE_ERR_OK) {
        WasmBotsMessage_PresentCircumstances_Destroy(circumstances);
        WasmBotsMessage__Error* errMsg = WasmBotsMessage__Error_Create();
        errMsg->description = "Could not read PresentCircumstances message";
        errMsg->description_len = strlen(errMsg->description);
        wsmbt_logErr(errMsg->description);
        playerMove = errMsg;
    }
    else {
        playerMove = _clientTick(circumstances);
    }

    if (playerMove != NULL) {
        _dataAccess.position = 0;
        err = WasmBotsMessage_WriteBytes(&_dataAccess, playerMove, true);
        err = WasmBotsMessage_Destroy(playerMove);
    }

    WasmBotsMessage_PresentCircumstances_Destroy(circumstances);
}
//// \SETUP AND LOOP

File: libraries/Go/wasmbot_client/wasmbot_client.go
package wasmbot_client

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"unsafe"
)

const GP_VERSION uint16 = 7
const MAX_NAME_LEN int = 26

type BotMetadata struct {
	Name    string
	Version [3]uint16
}

func Log(msg string) {
	msgBytes := []byte(msg)
	HostLogFunction(2, uintptr(unsafe.Pointer(&msgBytes[0])), uintptr(len(msgBytes)))
}

func LogErr(msg string) {
	msgBytes := []byte(msg)
	HostLogFunction(0, uintptr(unsafe.Pointer(&msgBytes[0])), uintptr(len(msgBytes)))
}

type ClientSetupFunction func() BotMetadata
type ClientReceiveGameParamsFunction func(InitialParameters) bool
type TickFunction func(PresentCircumstances) Message

var clientSetup ClientSetupFunction = _clientSetupNoop
var clientReceiveGameParams ClientReceiveGameParamsFunction = _clientReceiveGameParamsNoop
var clientTick TickFunction = _clientTickNoop

func _clientSetupNoop() BotMetadata {
	LogErr("No ClientSetupFunction set!")
	return BotMetadata{Name: "[INVALID]", Version: [3]uint16{0, 0, 0}}
}

func _clientReceiveGameParamsNoop(params InitialParameters) bool {
	LogErr("No ClientReceiveGameParamsFunction set!")
	return true
}

func _clientTickNoop(_ PresentCircumstances) Message {
	errMsg := _Error{Description: "No client tick function registered"}
	return errMsg
}

func RegisterClientSetup(setup ClientSetupFunction) {
	clientSetup = setup
}

func RegisterClientReceiveGameParams(receiveGameParams ClientReceiveGameParamsFunction) {
	clientReceiveGameParams = receiveGameParams
}

func RegisterTickCallback(cb TickFunction) {
	clientTick = cb
}

var hostReserve []byte

func reserveHostMemory(size uintptr) bool {
	if hostReserve != nil {
		LogErr("CLIENT ERROR: Attempting to reserve memory twice")
		return false
	}
	hostReserve = make([]byte, size)
	return true
}

//export setup
func Setup(request uintptr) uintptr {
	reserveHostMemory(request)

	botData := clientSetup()

	offset := 0
	copy(hostReserve[offset:], []byte(botData.Name))
	offset += len(botData.Name)
	paddingSize := MAX_NAME_LEN - len(botData.Name)
	if paddingSize > 0 {
		copy(hostReserve[offset:offset+paddingSize], make([]byte, paddingSize))
	}
	offset = MAX_NAME_LEN
	for vei, ve := range botData.Version {
		binary.LittleEndian.PutUint16(hostReserve[offset+vei*2:], ve)
	}
	offset += len(botData.Version) * 2

	return uintptr(unsafe.Pointer(&hostReserve[0]))
}

//export receiveGameParams
func ReceiveGameParams(offset uintptr) bool {
	paramsReader := bytes.NewReader(hostReserve[offset:])
	initParams, err := InitialParametersFromBytes(paramsReader)
	if err != nil {
		LogErr(fmt.Errorf("ERROR: Could not parse InitialParameters in host reserve: %w", err).Error())
		return false
	}
	if initParams.ParamsVersion != GP_VERSION {
		LogErr(fmt.Sprintf("ERROR: Can't parse GameParams v%d; only prepared for v%d", initParams.ParamsVersion, GP_VERSION))
		return false
	}

	return clientReceiveGameParams(*initParams)
}

// not sure if this is more broadly useful...
type sliceWriter struct {
	data   []byte
	offset int
}

func (w *sliceWriter) Write(p []byte) (n int, err error) {
	if w.offset+len(p) > len(w.data) {
		return 0, io.ErrShortWrite
	}
	n = copy(w.data[w.offset:], p)
	w.offset += n
	return n, nil
}

//export tick
func Tick(offset uintptr) {
	circsReader := bytes.NewReader(hostReserve[offset:])
	circumstances, err := PresentCircumstancesFromBytes(circsReader)

	var submittedMove Message
	if err != nil {
		errStr := fmt.Errorf("ERROR: Could not parse PresentCircumstances in host reserve: %w", err)
		LogErr(errStr.Error())
		submittedMove = _Error{Description: errStr.Error()}
	} else {
		submittedMove = clientTick(*circumstances)
	}

	moveWriter := &sliceWriter{data: hostReserve, offset: 0}
	submittedMove.WriteBytes(moveWriter, true)
}
